aries_cloudagent/messaging/valid.py

```diff
@@ -219,6 +239,28 @@ def __init__(self):
         )
 
 
+class SDJSONWebToken(Regexp):
+    """Validate SD-JSON Web Token."""
+
+    EXAMPLE = (
+        "eyJhbGciOiJFZERTQSJ9."
+        "eyJhIjogIjAifQ."
+        "dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk"
+        "~WyJEM3BUSFdCYWNRcFdpREc2TWZKLUZnIiwgIkRFIl0"
+        "~WyJPMTFySVRjRTdHcXExYW9oRkd0aDh3IiwgIlNBIl0"
+        "~WyJkVmEzX1JlTGNsWTU0R1FHZm5oWlRnIiwgInVwZGF0ZWRfYXQiLCAxNTcwMDAwMDAwXQ"
+    )
+    PATTERN = r"^[-_a-zA-Z0-9]*\.[-_a-zA-Z0-9]*\.[-_a-zA-Z0-9]*~[a-zA-Z0-9~]*$"
```

Comment:
Sorry to keep on coming back to this issue :grimacing: :sweat_smile:

This pattern will match SD-JWTs fine, I think, but it can also match some other unexpected strings. For instance, it could match:

```
eyJhbGciOiJFZERTQSJ9.eyJhIjogIjAifQ.dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk~~~~~~~~~~~~~~
```
For that matter, it can also match:

```
...~
```
And
```
...~~~~~~
```

So we need to be careful about inefficient regex like that catastrophic backtracking but we still want to impose some structure and make sure there are values between the delimiters. At least superficially. We don't need to go too crazy, of course.

My suggestion (hopefully it's a valid suggestion this time :sweat_smile: ):
```suggestion
    PATTERN = r"^[-_a-zA-Z0-9]+\.[-_a-zA-Z0-9]+\.[-_a-zA-Z0-9]+(?:~[a-zA-Z0-9]+)*~?$"
```
The final chunk is a non-capturing group that  requires text to follow but a trailing `~` is permitted to enable the disclosures without the key binding to be expressed (I think I'm remembering that correctly)

Reply:
Can encoded disclosures have a `_` in them?